aten/src/TH/generic/THTensorMath.c

```diff
@@ -3482,106 +3482,123 @@ void THTensor_(bhistc)(THTensor *hist, THTensor *tensor, int64_t nbins, real min
-                                THTensor_(digamma_one)(alpha)) * const_cdf
-                             + -a0 * a0 + b1 * x * (
-                               +a1 * a1 + b2 * x / 2 * (
-                               -a2 * a2 + b3 * x / 3 * (
-                               +a2 * a3 + b4 * x / 4 * (
-                               -a4))));
-  const real result = -const_cdf_alpha * one_over_const_pdf;
+  const real factor = THTensor_(digamma_one)(alpha)
+                    - THTensor_(digamma_one)(alpha + beta) - TH_MATH_NAME(log)(x);
+  real numer = 1;
+  real series = numer / alpha * (factor + 1 / alpha);
+  for (int i = 1; i <= 9; ++i) {
+    numer *= (i - beta) * x / i;
+    const real denom = alpha + i;
+    series += numer / denom * (factor + 1 / denom);
+  }
+  const real result = x * TH_MATH_NAME(pow)(1 - x, 1 - beta) * series;
   return isnan(result) ? 0.0 : result;
 }
 
 // Approximate reparameterized gradient of Beta(x,alpha,beta) wrt beta.
-// Assumes x is close to zero.
+// Assumes x is close to zero and uses a Taylor expansion.
 static inline real THTensor_(beta_grad_beta_small)(real x, real alpha, real beta) {
-  const real a0 = 1 / alpha;
-  const real a1 = 1 / (alpha + 1);
-  const real a2 = 1 / (alpha + 2);
-  const real a3 = 1 / (alpha + 3);
-  // Let pdf = pow(x,alpha-1) * pow(1-x,beta-1) / Beta(alpha,beta).
-  // Let const = Beta(alpha,beta) / pow(x, alpha). Then
-  const real one_over_const_pdf = x / TH_MATH_NAME(pow)(1 - x, beta - 1);
-  const real const_cdf = +a0 + (beta - 1) * x * (
-                         -a1 + (beta - 2) * x / 2 * (
-                         +a2 + (beta - 3) * x / 3 * (
-                         -a3)));
-  const real const_cdf_beta = (THTensor_(digamma_one)(alpha + beta) -
-                               THTensor_(digamma_one)(beta)) * const_cdf
-                            + 0 + x * (
-                            -a1 + x / 2 * (
-                            +a2 * (2 * beta - 3) + x / 3 * (
-                            -a3 * (3 * beta * beta - 12 * beta + 11))));
-  const real result = -const_cdf_beta * one_over_const_pdf;
+  const real factor = THTensor_(digamma_one)(alpha+beta) - THTensor_(digamma_one)(beta);
+  real numer = 1;
+  real betas = 1;
+  real dbetas = 0;
+  real series = factor / alpha;
+  for (int i = 1; i <= 8; ++i) {
+    numer *= -x / i;
+    dbetas = dbetas * (beta - i) + betas;
+    betas = betas * (beta - i);
+    series += numer / (alpha + i) * (dbetas + factor * betas);
+  }
+  const real result = -x * TH_MATH_NAME(pow)(1 - x, 1 - beta) * series;
   return isnan(result) ? 0.0 : result;
 }
 
+// Approximate reparameterized gradient of Beta(x,alpha,beta) wrt alpha.
+// Assumes alpha and beta are both large and uses a Rice saddle point expansion.
+// To ensure numerical stability, this computation is performed at higher precision.
+static inline real THTensor_(beta_grad_alpha_mid)(double x, double alpha, double beta) {
```

Comment:
@martinjankowiak If you have time, could you try to make this computation more numerically stable so we can use single precision?